Stock.Api/Controllers/ProviderController.cs

```diff
@@ -0,0 +1,145 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using AutoMapper;
+using Microsoft.AspNetCore.Mvc;
+using Stock.Api.DTOs;
+using Stock.Api.Extensions;
+using Stock.AppService.Services;
+using Stock.Model.Entities;
+
+namespace Stock.Api.Controllers
+{
+    [Produces("application/json")]
+    [Route("api/provider")]
+    [ApiController]
+    public class ProviderController : ControllerBase
+    {
+        private ProviderService service;
+        private readonly IMapper mapper;
+
+        public ProviderController(ProviderService service, IMapper mapper)
+        {
+            this.service = service;
+            this.mapper = mapper;
+        }
+
+        /// <summary>
+        /// Permite crear un Proveedor
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        [HttpPost]
+        public ActionResult Post([FromBody] ProviderDTO value)
+        {
+            TryValidateModel(value);
+
+            try
+            {
+                var provider = this.mapper.Map<Provider>(value);
+                this.service.Create(provider);
+                value.Id = provider.Id;
+                return Ok(new { Success = true, Message = "", data = value });
+            }
+            catch
+            {
+                return Ok(new { Success = false, Message = "The name is already in use" });
+            }
+        }
+
+        /// <summary>
+        /// Lista los proveedores existentes 
+        /// </summary>
+        /// <returns></returns>
+        [HttpGet]
+        public ActionResult<IEnumerable<ProviderDTO>> Get()
+        {
+            try
+            {
+                var result = this.service.GetAll();
+                return this.mapper.Map<IEnumerable<ProviderDTO>>(result).ToList();
+            }
+            catch (Exception)
+            {
+                return StatusCode(500);
+            }
+        }
+        
+
+        /// <summary>
+        /// Permite consultar un proveedor por su Id
+        /// </summary>
+        /// <param name="id"></param>
+        /// <returns></returns>
+        [HttpGet("{id}")]
+        public ActionResult<ProviderDTO> Get(string id)
+        {
+            try
+            {
+                var result = this.service.Get(id);
+                return this.mapper.Map<ProviderDTO>(result);
+            }
+            catch (Exception)
```

Comment:
No está mal atrapar excepciones (obvio, está bien), pero:

- ¿por qué algunos métodos tienen y otros no? (como el de abajo)
- ¿Por qué atraparías excepciones default (sin nombrarlas o sin tipar)?
- ¿hay algún bad smell que estes viendo?
- ¿como lo arreglarías?

Reply:
Si la idea es que todos los métodos puedan atrapar las excepciones y hacer las respectivas acciones  de acuerdo al tipo de  excepción que se presente, por ello es importante nombrar las excepciones ,  y no simplemente agarrar la excepción generica ya que No permite hacer una acción de remediación y solo me permitiría in-visibilizar un mal comportamiento de la aplicación de cara al usuario. 
El Bad Smell que se presenta es  "Dispensable - Código Duplicado" si realizo el manejo de excepciones en cada método, la forma de arreglarlo es crear una clase para el manejo de excepciones.